### PR TITLE
Add tic-tac-toe game

### PR DESCRIPTION
**Pull request by el-ethan**
Add tic-tac-toe game.

### CODE REVIEW

File: next-env.d.ts [hunk 1]
```diff
@@ -1,3 +1,6 @@
 /// <reference types="next" />
 /// <reference types="next/types/global" />
 /// <reference types="next/image-types/global" />
+
+// NOTE: This file should not be edited
+// see https://nextjs.org/docs/basic-features/typescript for more information.
```
Comment (el-ethan) [inline next-env.d.ts:6]: These changes were automatically generated by the new version of next.js. I think they can be ignored.

File: package.json [hunk 2]
```diff
@@ -9,17 +9,19 @@
...
   },
   "devDependencies": {
-    "@types/styled-components": "^5.1.11",
     "@types/react": "17.0.14",
+    "@types/styled-components": "^5.1.11",
+    "@typescript-eslint/parser": "^3.0.0",
     "eslint": "7.30.0",
     "eslint-config-next": "11.0.1",
     "prettier": "^2.3.2",
+    "prettier-eslint": "10.1.0",
```
Comment (el-ethan) [inline package.json:24]: The dependencies added here were an attempt to get prettier eslint to work. I haven't gotten it totally working yet.

File: pages/tic-tac-toe.tsx [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+const ClickedXCell = () => {
+    return <XCell>X</XCell>
+}
+
+const ClickedOCell = () => {
+    return <OCell>O</OCell>
+}
+
+const winningPatterns = [
```
Comment (el-ethan) [inline pages/tic-tac-toe.tsx:34]: The patterns represent the indexes of the cells in 3x3 grid that need to be filled to win.

File: pages/tic-tac-toe.tsx [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+  const [remainingChoices, setRemainingChoices] = useState(cells);
+
+  const [oIndexes, setOIndexes] = useState<number[]>([]);
+  const [xIndexes, setXIndexes] = useState<number[]>([]);
+  const [yourTurn, setYourTurn] = useState(true)
+
+  const findMoveForRobot = (yourChoices: number[]): number | undefined => {
+    for (let i = 0; i < winningPatterns.length; i++) {
+        const commonNumbers = winningPatterns[i].filter(num => yourChoices.includes(num))
```
Comment (el-ethan) [inline pages/tic-tac-toe.tsx:57]: Find how many numbers the player has in common with the winning any winning pattern.

File: pages/tic-tac-toe.tsx [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        let choice;
+        if (commonNumbers.length === 2) {
```
Comment (el-ethan) [inline pages/tic-tac-toe.tsx:59]: If the player has two numbers in common, then choose the third number in the winning pattern.

File: pages/tic-tac-toe.tsx [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+            return choice
+        } else {
+            continue
+        }
+    }
+  }
+
+  const checkWinner = (choices: number[]) => {
+    return winningPatterns.some(pattern => {
+      return pattern.filter(num => choices.includes(num)).length === 3
```
Comment (el-ethan) [inline pages/tic-tac-toe.tsx:73]: loop through all the winning patterns and find if you have three in common with the pattern.

File: pages/tic-tac-toe.tsx [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    })
+  }
+
+  const switchTurn = (index: number) => {
+      if (!yourTurn) {
```
Comment (el-ethan) [inline pages/tic-tac-toe.tsx:78]: Don't allow the rest of the click handler to run if it is not your turn.

File: pages/tic-tac-toe.tsx [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+      const youWon = checkWinner(allMyChoices)
+
+      if (youWon) {
+        checkAnswer(LOCK_LETTER, true, router)
+        return;
+      }
+
+      const choicesMinusMine = remainingChoices.filter(item => item !== index)
+      if (choicesMinusMine.length === 0) {
```
Comment (el-ethan) [inline pages/tic-tac-toe.tsx:93]: no more choices left.

File: pages/tic-tac-toe.tsx [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        checkAnswer(LOCK_LETTER, false, router)
+        router.reload();
```
Comment (el-ethan) [inline pages/tic-tac-toe.tsx:95]: Start the game over.

File: pages/tic-tac-toe.tsx [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+      }
+
+      setRemainingChoices(choicesMinusMine)
+
+      setTimeout(() => {
```
Comment (el-ethan) [inline pages/tic-tac-toe.tsx:100]: The robot's turn is done after a 2 second delay, just to give the feeling that the robot is "thinking"

File: .github/workflows/tests.yml [hunk 11]
```diff
@@ -0,0 +1,19 @@
+name: Tests and Linting
```
Comment (el-ethan) [inline .github/workflows/tests.yml:1]: Github action to get tests and linting running on every PR

File: pages/tic-tac-toe/index.tsx [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+            item => item !== index
+        );
+        if (choicesMinusMine.length === 0) {
+            checkAnswer(LOCK_LETTER, false, router);
+            router.reload();
+        }
+
+        setRemainingChoices(choicesMinusMine);
+
+        setTimeout(() => {
```
Comment (lbskinner) [inline pages/tic-tac-toe/index.tsx:111]: Just curious, why setTimeout here?

**Review by lbskinner — APPROVED** [review]

File: pages/tic-tac-toe/index.tsx (hunk 12 above)
Comment (el-ethan) [reply pages/tic-tac-toe/index.tsx:111]: This is just to give the appearance that the robot is "thinking" before making its move.

**Pull request merged**